### ISSUE TITLE
transaction with mutex lock

### ISSUE THREAD

**Issue by RoarGronmo** [simolus3/moor#2167]
When using transactions, in different simtainously asyncronious parts of my code, I may come in this situation:

```
I/flutter ( 9897): ----------------FIREBASE CRASHLYTICS----------------
I/flutter ( 9897): SqliteException(1): cannot start a transaction within a transaction
I/flutter ( 9897):   Causing statement: BEGIN TRANSACTION
I/flutter ( 9897): #0      DatabaseImpl.execute (package:sqlite3/src/ffi/impl/database.dart:134:9)
I/flutter ( 9897): #1      Sqlite3Delegate.runWithArgsSync (package:drift/src/sqlite3/database.dart:107:16)
I/flutter ( 9897): #2      _NativeDelegate.runCustom.<anonymous closure> (package:drift/native.dart:171:30)
I/flutter ( 9897): #3      new Future.sync (dart:async/future.dart:302:31)
I/flutter ( 9897): #4      _NativeDelegate.runCustom (package:drift/native.dart:171:19)
I/flutter ( 9897): #5      _BaseExecutor.runCustom.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:109:19)
I/flutter ( 9897): #6      _BaseExecutor._synchronized (package:drift/src/runtime/executor/helpers/engines.dart:55:20)
I/flutter ( 9897): #7      _BaseExecutor.runCustom (package:drift/src/runtime/executor/helpers/engines.dart:105:12)
I/flutter ( 9897): #8      _StatementBasedTransactionExecutor.ensureOpen.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:200:17)
I/flutter ( 9897): #9      _BaseExecutor._synchronized (package:drift/src/runtime/executor/helpers/engines.dart:55:20)
I/flutter ( 9897): #10     _StatementBasedTransactionExecutor.ensureOpen (package:drift/src/runtime/execu
I/flutter ( 9897): ----------------------------------------------------
I/TRuntime.CctTransportBackend( 9897): Making request to: https://crashlyticsreports-pa.googleapis.com/v1/firelog/legacy/batchlog
I/flutter ( 9897): ----------------FIREBASE CRASHLYTICS----------------
I/flutter ( 9897): CouldNotRollBackException: SqliteException(1): cannot rollback - no transaction is active
I/flutter ( 9897):   Causing statement: ROLLBACK TRANSACTION. 
I/flutter ( 9897): For context: The transaction was rolled back because of SqliteException(1): cannot commit - no transaction is active
I/flutter ( 9897):   Causing statement: COMMIT TRANSACTION, which was thrown here: 
I/flutter ( 9897): #0      DatabaseImpl.execute (package:sqlite3/src/ffi/impl/database.dart:134:9)
I/flutter ( 9897): #1      Sqlite3Delegate.runWithArgsSync (package:drift/src/sqlite3/database.dart:107:16)
I/flutter ( 9897): #2      _NativeDelegate.runCustom.<anonymous closure> (package:drift/native.dart:171:30)
I/flutter ( 9897): #3      new Future.sync (dart:async/future.dart:302:31)
I/flutter ( 9897): #4      _NativeDelegate.runCustom (package:drift/native.dart:171:19)
I/flutter ( 9897): #5      _BaseExecutor.runCustom.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:109:19)
I/flutter ( 9897): #6      _BaseExecutor._synchronized (package:drift/src/runtime/executor/helpers/engines.dart:55:20)
I/flutter ( 9897): #7      _BaseExecutor.runCustom (package:drift/src/runtime/executor/helpers/engines.dart:105:12)
I/flutter ( 9897): #8      _S
I/flutter ( 9897): #0      _extension#1.rollbackAfterException (package:drift/src/runtime/api/connection_user.dart:578:7)
I/flutter ( 9897): <asynchronous suspension>
I/flutter ( 9897): #1      DatabaseConnectionUser.transaction.<anonymous closure>.<anonymous closure> (package:drift/src/runtime/api/connection_user.dart:470:15)
I/flutter ( 9897): <asynchronous suspension>
I/flutter ( 9897): #2      DatabaseConnectionUser.transaction (package:drift/src/runtime/api/connection_user.dart:449:12)
I/flutter ( 9897): <asynchronous suspension>
I/flutter ( 9897): ----------------------------------------------------

```
I am pretty sure that my transactions sections aren't nested, where a such situation may occure. 

```
Future<void> getOrderLines({
  required int? ordNo,
  String? ordHash,
  int? frmNo,
  String? frmHash,
  String? officeNo,
}) async {

  if(ordNo==null){
    Message.error(null,"No valid order number ($ordNo), could not reload order lines");
    return;
  }

  if(ordHash==null){
    var order = await globals.mWorkDatabase.getOrder(ordNo).getSingleOrNull();

    if(order==null){
      Message.error(null, "Could not find order $ordNo");
      return;
    }

    ordHash = order.hash;

    if(order.hash == null){
      Message.warn(null, "Could not find hash for order $ordNo");
      return;
    }
  }
  
  if(officeNo==null){
    var orderFirmsOffice = await globals.mWorkDatabase.getSelectedOffice2(SettingKey.userSelectionSelectedOffice).getSingleOrNull();
    officeNo = orderFirmsOffice?.rNo;
    if(officeNo==null) {
      Message.warn(null, "Office not selected ?");
    }
  }
  
  if((frmHash==null)||(frmNo==null)){
    var selectedFirm = await globals.mWorkDatabase.getSelectedFirm(
        SettingKey.userSelectionSelectedFirm,
        SettingKey.userSelectionSelectedCountry
    ).getSingleOrNull();
    frmHash = selectedFirm?.hash;
    frmNo = selectedFirm?.frmNo;
    if(frmHash == null){ 
      Message.warn(null, "Firm hash not found !");
      return;
    }
    if(frmNo == null){
      Message.error(null, "FirmNo not found !");
      return;
    }
  }




  await ApiService().getFirmsFrmNoOrdersOrdNoOrderLines(
      frmNo: frmNo, 
      ordNo: ordNo, 
      ordHash: ordHash
  )?.then((ordLines) {
    globals.mWorkDatabase.transaction(() async{    //<-- First transaction
      if(ordLines!=null){
        for(var element in ordLines){
          if(element!=null){
            await globals.mWorkDatabase.into(globals.mWorkDatabase.orderLinesTable).insertOnConflictUpdate(
                OrderLinesTableCompanion(
                    readTime: drift.Value(DateTime.now().millisecondsSinceEpoch),
                    frmNo: drift.Value(frmNo??0),
                    frmHash: drift.Value(frmHash),
                    ordNo: drift.Value(element.ordNo??0),
                    lnNo: drift.Value(element.lnNo??0),
                    hash: drift.Value(element.hash),
                    r1: drift.Value(element.r1??0),
                    empNo: drift.Value(element.empNo??0),
                    empHash: drift.Value(element.empHash),
                    rsp: drift.Value(element.rsp??0),
                    un: drift.Value(element.un??0),
                    delDt: drift.Value(element.delDt??0),
                    cfDelDt: drift.Value(element.cfDelDt??0),
                    durDt: drift.Value(element.durDt??0),
                    r10: drift.Value(officeNo),
                    prodNo: drift.Value(element.prodNo),
                    descr: drift.Value(element.descr),
                    noteNm: drift.Value(element.noteNm),
                    noteHash: drift.Value(element.noteHash)
                )
            );  
          }
        }
      }

    });

    //Get the products
    if(ordLines!=null){
      globals.mWorkDatabase.transaction(() async {  //<-- Second transaction
        for(var ordLine in ordLines){
          if(ordLine?.ordNo!=null) {
            //Check if product already exists in table

            var currentProduct = await globals.mWorkDatabase.getProduct(SettingKey.userSelectionSelectedFirm, ordLine?.prodNo??"").getSingleOrNull();

            //Load it if it doesn't exist
            if(currentProduct==null) {
              await ApiService().getFirmsProduct(
                  frmNo: frmNo,
                  frmHash: frmHash,
                  prodNo: ordLine?.prodNo
              )?.then((product) async {
                print("L1090: frmHash:$frmHash prodNo=${ordLine
                    ?.prodNo}: $product");
                if (product != null) {
                  await globals.mWorkDatabase.into(
                      globals.mWorkDatabase.orderFirmsProductsTable)
                      .insertOnConflictUpdate(
                      OrderFirmsProductsTableCompanion(
                          readTime: drift.Value(DateTime
                              .now()
                              .millisecondsSinceEpoch),
                          prodNo: drift.Value(product.prodNo ?? ""),
                          frmNo: drift.Value(frmNo ?? 0),
                          descr: drift.Value(product.descr),
                          prodTp3: drift.Value(product.prodTp3 ?? 0),
                          prodPrG2: drift.Value(product.prodPrG2 ?? 0),
                          prodGr: drift.Value(product.prodGr ?? 0),
                          stSaleUn: drift.Value(product.stSaleUn ?? 0),
                          webPg2: drift.Value(product.webPg2),
                          r3: drift.Value(product.r3 ?? 0),
                          r4: drift.Value(product.r4 ?? 0)
                      )
                  );
                }
              });
            }
          }
        }
      });
    }


  });
  

  

}


```

(I know my code is cruel...)

The fact is that the first transaction i started with an iteration inside which executes over a long time and the second transaction starts asyncroniously before the first has finished.

You should add some kind of "mutex" option on the transactions, or build some kind of handler that could organize multiple transactions at the same time in an asynchrone scenario.

**Comment by Mike278** [comment]
You're not `await`ing each transaction so I'm guessing that will fix the issue, but Drift does implement a mutex around transactions so I'm a bit surprised that code caused that error.

**Comment by simolus3** [comment]
I think this has been fixed with f6830bcd68afecfd51aafc1462da796f14be1677. @RoarGronmo, could you add this to your pubspec and let me know if it fixes this problem? If everything works, I'll release a new drift version including this fix this weekend:

```yaml
dependency_overrides:
  drift:
    git:
      url: https://github.com/simolus3/drift.git
      ref: develop
      path: drift
```

**Comment by RoarGronmo** [comment]
@Mike278 Of course I could, but then I expect the call to be janky... but since it is inside an async call you may have right.

@simolus3 I'll do a try, give me some minutes, and I'll report back to you (I actually changed my code in this situatuin to cope with that error, I need to deep dive in my repo's...)

**Comment by RoarGronmo** [comment]
@simolus3 Naah: Got this error:

```
PS C:\FlutterProjects\mwork> flutter pub get
Running "flutter pub get" in mwork...                            2 858ms
Resolving dependencies...
Because no versions of drift_dev match >2.2.0+1 <3.0.0 and drift_dev >=2.1.0 <2.2.0+1 depends on drift >=2.0.0 <2.2.0, drift_dev >=2.1.0 <2.2.0+1 or >2.2.0+1 <3.0.0 requires drift >=2.0.0 <2.2.0.
And because drift_dev 2.2.0+1 depends on drift >=2.0.0 <2.3.0, drift_dev ^2.1.0 requires drift from hosted.
So, because mwork depends on both drift from git and drift_dev ^2.1.0, version solving failed.
pub get failed
command: "c:\Flutter\flutter\bin\cache\dart-sdk\bin\dart __deprecated_pub --color --directory . get --example"
pub env: {
  "FLUTTER_ROOT": "c:\Flutter\flutter",
  "PUB_ENVIRONMENT": "flutter_cli:get",
  "PUB_CACHE": "C:\Users\RoarGronmo\AppData\Local\Pub\Cache",
}
exit code: 1
last line of pub output: "So, because mwork depends on both drift from git and drift_dev ^2.1.0, version solving failed."

```
Do i need to do the same with `drift_dev` section ?

Secondly I see that other dependent libraries such as `drift_db_viewer: ^2.0.0` might have problems too when changing `drift` dependency scheme.

**Comment by simolus3** [comment]
Right, you probably have to [override every drift package](https://drift.simonbinder.eu/docs/internals/#using-an-unreleased-drift-version). Third-party packages like `drift_db_viewer` shouldn't be a problem though, drift's public API hasn't changed.

**Comment by kuhnroyal** [comment]
I am also seeing this problem but can not test this because of #2185

**Comment by kuhnroyal** [comment]
I am using `await Future.wait(multipleInserts)`.

**Comment by kuhnroyal** [comment]
Works with 2.3.0 after restructuring my migrations.

**Comment by simolus3** [comment]
I think this should have been fixed in 2.3.0 - the migration issues were a separate issue.

If you're still running into this with drift 2.3.0, please let me know - I'll reopen the issue and take another look then.
